xbrl_parser/helper/connection_manager.py

```diff
@@ -0,0 +1,60 @@
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+import requests
+import time
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class ConnectionManager:
+    """
+    This class handles http requests. If a requests fails, the method will retry the http requests as many times as specified.
+    Between each requests the process will sleep {backoff factor} * (2 ** ({number of total retries} - 1)) seconds between failed requests.
+    After the request is successfull, the process will sleep delay / 1000 seconds so that it is compliant with sec's policy.
+
+    https://www.sec.gov/privacy.htm#security
+
+    """
+
+    def __init__(self, delay: int = 500, retries: int = 5, backoff_factor: float = 0.8, headers: dict = None):
+        """
+
+        @param from_locator: Specifies sleeping time after the request is successfull.
+        @param retries: How many times a request will be tried before assuming its failure.
+        @param backoff_factor: Used to measure time to sleep between failed requests.
+            The formula used is {backoff factor} * (2 ** ({number of total retries} - 1))
+        @param headers: Headers to use in http request.
+        """
+        self._delay = delay
+        self._retries = retries
+        self._backoff_factor = backoff_factor
+        self._headers = headers
+        self._session = self._create_session()
+
+    def download(self, url: str, headers: str):
+
+        response = self._session.get(url, headers=headers, allow_redirects=True)
+        logger.info(str(response.status_code) + " " + url)
+        # Set a timeout, so that we do not get blocked by the for making to many requests
+        time.sleep(self._delay / 1000)
+
+        return response
+
+    def _create_session(self,
+                        status_forcelist: tuple = (500, 502, 504, 403)) -> requests.Session:
```

Comment:
I think i would add the status 503 (Service Unavailable) by default to the status_forcelist.

Additionally I would remove 403 (Forbitten) from the default status_forcelist because normally if the server sends a 403 that means that the client will not get access even after trying multiple times.
As stated by [IETF in HTTP/1.1:](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.3):
"The client should not automatically repeat the request with the same credentials."

What do you think ?

Reply:
I agree we could add 503 to the status_forcelist. 

However I would not remove 403. SEC is not compliant with the standards and returns a 403 ERROR if you exceed their rate limit as explained here: [https://developer.edgar-online.com/docs/errors](https://developer.edgar-online.com/docs/errors).

From my experience, if you get a 403 ERROR but retry it some time afterwards, you usually get a 200 response the second time.

If SEC changes its 403 rate limit error code to 429, I believe it then should be changed as you suggest.

What do you think?